### DISCUSSION TITLE
App crashes on macOS 15.0 Sequoia - v4.3.2 fixes the issue!

### DISCUSSION THREAD

**Discussion by PTriebold** [MonitorControl/MonitorControl#1634] [General]
**Update 2/10/2024:** The new v4.3.2 release fixes all reported Sequoia related issues.

---

_Original issue as follows:_

### Before opening the issue, have you...?

- [X] Searched for existing issues
- [X] Looked through [the wiki](https://github.com/MonitorControl/MonitorControl/wiki)
- [X] Updated MonitorControl to the latest version (if applicable)

### Describe the bug

In the new Sequioa beta, MonitorControl will crash when setting the brightness of the internal monitor (16" Macbook). External Monitors work as expected.

### Steps to reproduce

1. Update to MacOs 15 Beta
2. Change internal brightness trough MonitorControl

### Expected behavior

The app changes the brightness of the internal display without crashing

### Anything else?

I've tested it with a fresh install of monitorcontrol.  
It actually changes the brightness of the internal monitor. If you disable the soft transition, it will be set to whatever was selected. It will get close otherwise.  
Reseting the internal display also crashes monitorcontrol.

### Environment Information (please complete the following information)

```markdown
- macOS version: 15.0 Sequoia (24A5264n)
- Mac model: Macbook Pro (16-Inch 2021)
- MonitorControl version: 4.2.0
- Monitor(s): Internal, HP Z27k G3, HP Z27k G3
- Apple Silicon/M1 (yes or no): yes
```

**Comment by waydabber** [comment]
Here is a build with macOS Sequoia support:

https://github.com/waydabber/MonitorControl/releases/tag/v4.3.0-beta

**Comment by alextuan1024** [comment]
Thank you for your contribution. 
Does the volume control work fine on your monitor?
It's not working on my LG.

**Comment by waydabber** [comment]
It should generally work, yes. If it doesn't, it might be a monitor issue. Or did it work before and not on Sequoia?

**Comment by alextuan1024** [comment]
Yeah it worked before, when the version was Sonoma. 
My best guess is the protocols are not very well supported on Sequoia.

**Comment by waydabber** [comment]
Are you on Intel or Apple Silicon? DDC should work fine (found no issues during tests). Does brightness control work?

**Comment by alextuan1024** [comment]
My device is a MacBook Pro with an M3 Pro chip, connected to an LG 27UP850 monitor, which also works as a power adapter via USB-C. 
Yes, the brightness control works fine. 
And I just found that the volume control works with the laptop cover open. But the volume slider disappears when I close the cover and only use the external display. 
If you need logs to debug, please lemme know.

**Comment by alextuan1024** [comment]
Not sure if this screenshot helps:
<img width="842" alt="image" src="https://github.com/MonitorControl/MonitorControl/assets/13645163/301758b5-4017-49a6-bb0e-ab135ec3241f">

**Comment by waydabber** [comment]
Oh, all right, the problem is that when the lid is closed, your display is set to be in HDR mode. Sequoia made a change in HDR mode that confuses MonitorControl - the app treats it as an Apple display as it confuses Seqoia's new SDR peak brightness adjustment slider with a native brightness control which up to this point was only available for Apple (and some Apple-LG) displays and therefore turns DDC off.

It is probably not feasible to fix this in MonitorControl 4.x as a more fundamental change would be needed, which would kick-off a chain reaction me starting to rewrite/update the entire app due to its generally outdated nature... You can instead use the latest version of **[BetterDisplay](https://github.com/waydabber/BetterDisplay)** in which this issue is already elegantly resolved + has all the features of MonitorControl for free (no license required).

**Comment by alextuan1024** [comment]
Thank you for the prompt response. I'll try it out later.

**Comment by Roozenlz** [comment]
It does work! I'm using macOS Sequoia too.

**Comment by bgiesing** [comment]
No Issues enabled on the other repo so I’ll put it here

Seems the “General Menu Items Style” setting under App Menu got reversed, setting it to “Show as text” shows icons and “Show as icons” shows text.

https://github.com/user-attachments/assets/e4190f82-e710-46df-a230-7f11d656e006

**Comment by waydabber** [comment]
@bgiesing - Indeed, thanks for reporting - I'll fix this if I provide an updated unofficial version.

**Comment by cjbbb** [comment]
If the official version is released (For macos 14), can we just update to the official version from this beta version?

**Comment by waydabber** [comment]
No, this unofficial versions has auto-update removed so it won't auto-update. I might release a more proper unofficial version in the unofficial repo that can auto-update to future unofficial versions + fix the issue mentioned by [bgiesing](https://github.com/bgiesing).

**Comment by chrisburton** [comment]
> Here is a build with macOS Sequoia support:
> 
> https://github.com/waydabber/MonitorControl/releases/tag/v4.3.0-beta

App doesn't open for me on Intel.

**Comment by waydabber** [comment]
@chrisburton - indeed.

**Comment by waydabber** [comment]
@chrisburton - I posted an updated version (see the opening comment), please check whether this works. Also fixed the issue reported by @bgiesing.

**Comment by chrisburton** [comment]
> @chrisburton - I posted an updated version (see the opening comment), please check whether this works. Also fixed the issue reported by @bgiesing.

Works! Thank you

**Comment by Boshimanu** [comment]
Hi guys, disabling the HDR works ok, the thing that still does not work, it the keyboard volume and brightness. Before updating to the Sequoia, I was able, to control the brightness and volume from the keyboard, now it does not. Maybe the solution is the put a shortcut, and I tried, but I do not know how to do this in the app, I Intend in the "MonitorControl". Any help would be appreciated, thank you. The main problem is for the audio, for the brightness, I can adjusted from the bar.
Thanks again!

**Comment by Boshimanu** [comment]
> 

Solved IT. It was an authorisation problem. Thank You. Now it works! 🥰 Audio and brightness. THANK YOU 🙏🙏🙏

**Comment by mpowaa** [comment]
This beta version fixed the issue of adjusting the brightness level on my external display. 👍  

However, there is a bug that permanently removes the Dock when the Monitor Control app is closed.  I have to open Activity Monitor app and quit the Dock app for it to restart and reappear again.

**Comment by waydabber** [comment]
You mean the macOS Dock disappears? Not sure how that could happen. :) Doesn't seem to be doing it on my installation. Do you have a crash log of sorts?

**Comment by mpowaa** [comment]
Just check the Console crash reports but there are none.  Can you try this, open up Settings > Desktop & Dock > Turn on Automatically hide and show the Dock.  Then quit the Monitor Control app to see if your macOS Dock disappears permanently.

I found another way to bring it back is to toggle the Automatically hide and show the Dock option but that only brings back the Dock on the internal display.  Never shows up again on the external display.

**Comment by oerix** [comment]
Will there be an official RC with these fixes?

**Comment by waydabber** [comment]
@oerix - that's a great question, but I can't provide an answer right now. For the time being please use the provided version, it should work well.

**Comment by Boshimanu** [comment]
> @oerix - that's a great question, but I can't provide an answer right now. For the time being please use the provided version, it should work well.

I'm very pleased cause everything is working as before. It is not working the with the HDR active, but for me it's ok. Thank you so much for releasing this fix super Fast. Much apreciate. 🙏

**Comment by oerix** [comment]
> @oerix - that's a great question, but I can't provide an answer right now. For the time being please use the provided version, it should work well.

Thank you, I've just read your post [here](https://www.reddit.com/r/mac/comments/1e30j3h/comment/ld5g3q6). Appreciate you for provided the quick fix and your continued work. 👍🏼

**Comment by hustliweizeng** [comment]
thanks a lot

**Comment by krim** [comment]
It doesn't crash anymore, but it doesn't work on the external monitor when I use the brightness control keys. The brightness always changes on the laptop screen instead of external monitor. I tried changing the settings, but nothing changed.
![image](https://github.com/user-attachments/assets/a06f2e8e-8e49-444c-92f3-3b50d951dec0)

It only works when I use the slider from the tray.

**Comment by waydabber** [comment]
You need to configure Accessibility Permissions under System Settings. There might be an old MonitorControl there but the app has a different signature which might confuse things so you might need to remove the old one first.

**Comment by Boshimanu** [comment]
> You need to configure Accessibility Permissions under System Settings. There might be an old MonitorControl there but the app has a different signature which might confuse things so you might need to remove the old one first.

I confirm, happened to me, same thing 👍 
Remember to restart the app. 👍

**Comment by krim** [comment]
@waydabber thanks a lot. It works

**Comment by moresol** [comment]
> Here is a build with macOS Sequoia support:
> 
> https://github.com/waydabber/MonitorControl/releases/tag/v4.3.0-beta

Helpful, thank you.

**Comment by lianshuze** [comment]
When will the official version support MacOS Sequoia?

**Comment by JamesGeiger** [comment]
I appreciate the unofficial build, but will the fix ever be made official?

**Comment by waydabber** [comment]
Yes, I talked to the other project maintainers who are stepping down as active maintainers. I got the green light from them to go ahead and take over the project and do a proper official build with Sequoia compatibility. Auto-update from 4.2.0 to this new build will not work as the bundle identifier and signature will change going forward (the old one can't be used anymore due to technical issues, the new one will be same as the linked unofficial build...). Until then you can use the unofficial build, no reason for waiting to the official one - the only difference in an official build will be the re-enabled auto-update capability for future versions.

**Comment by JamesGeiger** [comment]
Ok great - thanks. I like your other app Better Display too btw. But I like the simplicity of this. Both are great apps

**Comment by ehehaye** [comment]
I' m using 4.3.1-beta1 on M1 Pro mac. The custom shortcuts for brightness and contrast works wrong.

<img width="730" alt="image" src="https://github.com/user-attachments/assets/62172bb1-beb7-43d7-9f92-5e9a28bf0baf">

**Comment by waydabber** [comment]
Indeed.

**Comment by moresol** [comment]
I use 4.3.0 version, everything is okay.

**Comment by Juiciator** [comment]
Hello the keyboard F1- F2 is not working on imac mini, any idea how to fix it? thanks

**Comment by waydabber** [comment]
@Juiciator - try fixing Accessibility Permissions for the app in System Settings.

**Comment by Juiciator** [comment]
> @Juiciator - try fixing Accessibility Permissions for the app in System Settings.

Thanks.. worked perfect 8-)

**Comment by gkndr** [comment]
thx a lot

**Comment by AllisonZw** [comment]
> @Juiciator - try fixing Accessibility Permissions for the app in System Settings.- 尝试在系统设置中修复应用程序的辅助功能权限。

nice 👍

**Comment by hmphu** [comment]
Thank you so much

**Comment by Fzx-oss** [comment]
In this version, do the F1 and F2 keys control the brightness by default?

**Comment by barathvelmu** [comment]
Its not opening at all on an Intel machine. How do I overcome this? I have already tried reinstalling, restarting etc.

**Comment by chrisburton** [comment]
@barathvelmu  What version?

**Comment by barathvelmu** [comment]
> @barathvelmu  What version?

v4.3.0, what is the link to v4.3.1? This fixed it for you?

**Comment by chrisburton** [comment]
@barathvelmu  Yes, version 4.3.1 fixed the issue. [Download 4.3.1 beta](https://github.com/waydabber/MonitorControl/releases/download/v4.3.1-beta/MonitorControl-v4.3.1-beta-macOS-Sequoia.zip).

**Comment by chrisburton** [comment]
> In this version, do the F1 and F2 keys control the brightness by default?

@Fzx-oss Yep. Keys F1 and F2 control the brightness.

**Comment by Fzx-oss** [comment]
@chrisburton 
Thank you for your quick response! I'm not sure if this is an error, but when I press F1 and F2 on the Mac's internal keyboard, they work as expected. However, when I use F1 and F2 on my external keyboard (HHKB), they don't function properly.

**Comment by supermebing** [comment]
nice 👍

**Comment by ghost** [comment]
Just wondering, what's preventing from the fix being submitted by PR and merged into the project?

**Comment by waydabber** [comment]
Nothing, I should just work on it. I actually started working on it but there is some issue with the keyboard shortcuts and I have zero idea what's wrong + there are various bugs and known issues (listed under Issues) that needs sorting out. I should invest more time, but I feel the code is so outdated that it does not makes no sense to start devoting time on sorting out all issues. But at the same time I am not inspired at all to do some half-baked official release either. So I now just decided to avoid the whole issue - for those who really want to continue using MonitorControl for now and can live with some of the issues the unofficial release should suffice. When I have time, I might make a MonitorControl 5 version derived from BetterDisplay (all excess features removed) but that should take a lot of time too - but at least then there would be a new and updated version.

I should probably make the unofficial release maybe "somewhat less unofficial" and link it on the opening page for easier discoverability for Sequoia users. Or maybe make it official but designate as beta or something.

**Comment by ghost** [comment]
Hey, thanks for your reply. I realize you've got the balance between BetterDisplay and MonitorControl. 
I was simply asking because I've come to love the granular volume control over DDC that MonitorControl provides and I'm unable to replicate using BetterDisplay.

**Comment by waydabber** [comment]
What do you mean by granular volume control? I tried to make sure I port all features of MonitorControl to BetterDisplay (these are available for free) because it would be cumbersome for users to run both apps at the same time if one needed this and that missing feature. You can change the volume control's min-max ranges, make fine scale default for keyboard control etc in both apps.

**Comment by chrisburton** [comment]
What about taking BetterDisplay and stripping it down to a minimal version and just package it as “BetterDisplay Lite” or something? Rather than trying to refactor MonitorControl and deal with all the bugs and headache. Just use your existing BD codebase.

**Comment by waydabber** [comment]
Yes, that's my plan B, but I'd create a MonitorControl 5.0 out of it instead. BetterDisplay has various diverging functions and making a BetterDisplay Lite focusing on brightness control would be confusing to those who associate the app with virtual screens or flexible resolutions, streaming, color mode selection, various tweaking options or some other capabilities.

**Comment by ghost** [comment]
> What do you mean by granular volume control? I tried to make sure I port all features of MonitorControl to BetterDisplay (these are available for free) because it would be cumbersome for users to run both apps at the same time if one needed this and that missing feature. You can change the volume control's min-max ranges, make fine scale default for keyboard control etc in both apps.

Thanks for your effort! I did try and use BetterDisplay. However I was unable to lock down the volume increments to 1. 
With the scale in BetterDisplay I had the values narrowed down to 3 and 4. However these increments somehow did not work consistently.

**Comment by waydabber** [comment]
@enviousjag - this is somewhat off-topic here, but one approach for finer control is to enable `Make fine scale default for volume keys`:

<img width="872" alt="Screenshot 2024-09-27 at 21 38 12" src="https://github.com/user-attachments/assets/96009fc6-3a3d-451f-a0c5-1b13099e8ab3">

This will change the default 16 steps to 64 (as the OSD has 16 steps, each with 4 substeps). However with the typcal scale DDC of 0-100 this will result in alternating steps of 1 and 2. So in order to make this work to change volume in 1 increments exactly, you need to change the volume scale to 0-64 (which should be fine for most displays) under `DDC Features...`:

<img width="872" alt="Screenshot 2024-09-27 at 21 40 25" src="https://github.com/user-attachments/assets/ca7d993c-603d-4aec-b243-f70c702be215">

MonitorControl indeed does things a bit differently - with fine scale it disregards the granularity of the macOS OSD scale and uses an exact increments of 1, however I think although this has benefits (and seemed good idea at the time) it has drawbacks as well, especially when one redefines the scale, so I choose to implement this feature differently in BetterDisplay.

Hope this helps!

**Comment by waydabber** [comment]
Hi all, posted an official v4.3.2 that should fix all issues:

- Sequoia compatibility
- No DDC control on third party displays in HDR mode
- Custom keyboard shortcuts not working
- Dock freezing on exit

Obviously the app could be improved in many ways still, the release was meant to simply make the app work on Sequoia as we all got used to it over the years. For more features, there is [BetterDisplay](https://betterdisplay.pro). :)
